compiler/rustc_ast/src/ast.rs

```diff
@@ -2848,11 +2848,11 @@ impl NormalAttr {
 }
 
 #[derive(Clone, Encodable, Decodable, Debug, HashStable_Generic)]
-pub struct AttrItem {
+pub struct Meta {
```

Comment:
I disagree with this renaming, "attribute items" and "meta items" are different things and shouldn't be mixed.

Meta items (`struct MetaItem`) are a restricted subset of attribute items, used for working with built-in attributes in the compiler (most of which fit into that subset) as an implementation detail.

The `meta` matcher in macros matches attr items, not meta items, but we cannot just rename it because it's user-facing and stable.

`NtMeta` could be renamed to `NtAttrItem` to better match its essence rather than surface syntax, but why bother, it's removed in https://github.com/rust-lang/rust/pull/124141 anyway.

Reply:
We could potentially do some larger naming reform to change `AttrItem` to `Meta`, and `MetaItem` to something like `RestrictedMeta`, but not as a part of a cleanup PR.

I'm actually not sure where the name "meta" came from (it is ancient) and whether it would make sense if we started naming things now.

Reply:
The basic point is that the current naming of types with the compiler is bad. Off the top of your head, do you know the distinction between `NormalAttr::tokens` and `AttrItem::tokens`? I sure don't, I have to look at the comments every time.

Looking at the [reference](https://doc.rust-lang.org/reference/attributes.html):

> InnerAttribute :
>    \# ! [ Attr ]
> 
> OuterAttribute :
>    \# [ Attr ]
> 
> Attr :
>    [SimplePath](https://doc.rust-lang.org/reference/paths.html#simple-paths) AttrInput?
> 
> AttrInput :
>       [DelimTokenTree](https://doc.rust-lang.org/reference/macros.html)
>    | = [Expression](https://doc.rust-lang.org/reference/expressions.html)

So:
- "Attribute" is the whole thing, including the leading `#` and (if present) `!`. That's fine.
- "Attr" is the part within the brackets. That's a bad name, much too similar to "Attribute", and when "Attribute" is inevitably abbreviated to "Attr" they'll be indistinguishable.
- "Attribute item" doesn't appear at all.
- "Meta item" is the restricted syntax used within the brackets for most builtin attributes. It's not a great name, but it's sufficiently different that it'll do.

The main problem is the lack of a good name for what the reference calls "Attr", i.e. the part within the brackets, not necessarily restricted to meta item syntax. That's the problem I'm trying to solve here. Something like "AttrContents" might work, and `NormalAttr::item` could become `NormalAttr::contents`. I want to avoid "item" as much as possible, it's too easily confused with normal [items](https://doc.rust-lang.org/reference/items.html).

Reply:
> Something like "AttrContents" might work

The [meta section](https://veykril.github.io/tlborm/decl-macros/minutiae/fragment-specifiers.html#meta) of TLBORM says "The `meta` fragment matches the contents of an attribute".

Reply:
I think there are two serious contenders here.
- "Attribute"/"Meta"/"Meta item": what this PR does. Pros: "Meta" matches the `meta` fragment specifier. Cons: "Meta" and "MetaItem" are subtly different, which is potentially confusing.
- "Attribute"/"AttrContents"/"Meta item". Pros: clearer terminology. Cons: doesn't match `meta` fragment specifier.

Reply:
I don't think the current naming is bad, the current scheme is as "strong contender" as those in the comment above, and "item" is a generic word for an element if a sequence that is clear from the context (like we don't mix "meta" in attributes and "meta" in metadata, for example).
I had `#[cfg_attr(pred, item1, item2, ..., item3)]` and the [`#[item1, item2, ..., item3]`](https://github.com/rust-lang/rfcs/pull/2600) proposal when coming up with the name. "Contents" doesn't fit here, but "item" does, and "meta" does too just because it's meaningless by itself.
I'd rather avoid churn here than start bikeshedding.

>Off the top of your head, do you know the distinction between `NormalAttr::tokens` and `AttrItem::tokens`?

This seems unrelated to the structure naming, but the answer is that I don't know and I'm not supposed to!
They return tokens in some form, lazy or not, the details I can lookup once in a year when I need them.
I do remember that the one from the trait returns lazy tokens exactly as they are kept in AST structures, so a mutable ref can be returned.

Reply:
I do think the naming is bad. It confuses me all the time. I didn't propose the name change just for fun, but because I found it made the code easier to understand. "Attribute item" is a name that doesn't come up in the reference, isn't sufficiently different to "Attribute", and overloads "item".

> I had #[cfg_attr(pred, item1, item2, ..., item3)] and the https://github.com/rust-lang/rfcs/pull/2600 proposal when coming up with the name. "Contents" doesn't fit here, but "item" does, and "meta" does too just because it's meaningless by itself.

The `AttrItem` type currently describes *everything* within the brackets. That doesn't line up with `item1`, `item2`, etc. A `meta` specifier matches everything in the brackets, but you rejected my renaming of `AttrItem` as `Meta`.

In `#[a, b, c]`, what word/phrase would you use to describe the `a, b, c` part?

Reply:
>but you rejected my renaming of `AttrItem` as `Meta`

I think it's fine as long as `MetaItem` is renamed to something else (https://github.com/rust-lang/rust/pull/127308#discussion_r1666491518), it's very generic but at least it will match the user-visible naming.

>In `#[a, b, c]`, what word/phrase would you use to describe the `a, b, c` part?

Attribute items, `Vec<AttrItem>`, as in the current naming.

>name that doesn't come up in the reference

That's probably the least convincing argument to me, the compiler code, RFCs and lang team are the source for terminology, and the reference follows, with a delay. If I weren't too lazy I could update the reference when doing the implementation, and it would say "attribute items" now.

Reply:
I can see I won't convince you. I will note that "meta item" occurs in lots of places in the compiler, including several error messages. "Attribute item" doesn't appear anywhere other than the `AttrItem` type.